### ISSUE TITLE
Column width not accurate

### ISSUE THREAD

**Issue by fbele** [exceljs/exceljs#744]
Hi,
If I want to define the column width, I need to add to almost every width a value of 0.71. So if I want a width of 12, I need actually to define it like `column.width = 12.71`.
Only when it comes to the width of less then 1, this addition actually varies.

Does anybody have an experience, what could cause such a behavior?

Thanks

**Comment by Siemienik** [comment]
@fbele may I please You to share with code?

**Comment by fbele** [comment]
Hi,
I just have a normal worksheet:

`const worksheet = workbook.addWorksheet('Sheet', {views: [{}]);`

And then I am adding my columns like following:

`
        worksheet.columns = [
            { key: 'A', width: 3.71 },
            { key: 'B', width: 5.42 },
            { key: 'C', width: 74.71 },
            { key: 'D', width: 22.71 },
            { key: 'E', width: 16.57 },
            { key: 'F', width: 12.71 },
            { key: 'G', width: 1 },
            { key: 'H', width: 22.71 },
            { key: 'I', width: 1 },
            { key: 'J', width: 50.71 }
        ];
`

The width values here have mostly (not always) added value of 0.71 value to the actual wanted width value.
For the width values that are less then 1, then I actually added a value of 0.42.
For the width above 74 value I added just 0.57 to it.
So this is not constant.

Here is an example of three columns in my project, which is the result of the width values configured above:

![image](https://user-images.githubusercontent.com/33025896/53045688-8a315500-348e-11e9-94a3-133530d5d762.png)

**Comment by jkRaccoon** [comment]
same on me. mine is Vuejs

**Comment by werkilla** [comment]
Thanks, I'm having the same problem in node with the stream workbook writer.  I thought my width values were just not taking but this helped me.

**Comment by yktakaha4** [comment]
The same thing happened in my environment (Excel for Mac, Version 16.41, In Japanese),
The solution was to multiply the width of each column by `1.17` instead of adding the value.

```
[...new Array(ws.columnCount)].map((_, index) => ws.getColumn(index + 1)).forEach((column) => column.width = (column.width ?? 0) * 1.17)
```

It seemed that the size of the width(幅) corresponding to the pixel(ピクセル) changed before and after saving by exceljs.
(Is there a phenomenon such as loss of column width units due to saving?)

[original]
![image](https://user-images.githubusercontent.com/20282867/94178278-dac42c00-fed5-11ea-9797-c93a181a5766.png)

[after saving(* 1.17)]
![image](https://user-images.githubusercontent.com/20282867/94178293-def04980-fed5-11ea-8ed2-cfbc310f528c.png)

[after saving(do nothing)]
![image](https://user-images.githubusercontent.com/20282867/94178557-368eb500-fed6-11ea-9603-b32ad26e0895.png)

**Comment by anterta** [comment]
no news?

**Comment by fbele** [comment]
> no news?

nope, unfortunately not. and as you see also not much attantion was put to.
it's also kind of baffeling that us few could experienced such a bug.
but back then I have given up on exceljs so I don't know if the bug was fixed in the newer releases.

**Comment by Siemienik** [comment]
I suppose it is dependent on which OS is used during opening a created file, or which spreadsheet program&version, or used system DPI, or default font.

1. Could I please about share with the wrong file with information witch OS and App you use?
2. Did you tested it during some different App or OS?

---
_I'm contributing ExcelJS just for a fun, If want you to know something more go into #1500_

**Comment by fbele** [comment]
> I suppose it is dependent on which OS is used while opening a created file, or which spreadsheet program&version, or used system DPI, or default font.
> 
> 1. Could I please share with the wrong file and information which OS and App you use?
> 2. Did you tested it during some different App or OS?
> 
> _I'm contributing ExcelJS just for fun. If want you to know something more go into #1500_

At that time I was using Angular on Windows 10 and I haven't thought of it testing it on some other app or OS. I didn't think that something like that could make a difference.
I hope I was able to answer your questions correctly.

**Comment by Siemienik** [comment]
I mean that, which software version are you using to open xlsx (for me it is Microsoft Macintosh Excel v16). It is essential because xlsx format keeps width as a character length, that occurs different width through different OS (dpi changed) or when different default font has been set. More information could be found on the MS Excel documentation page: https://docs.microsoft.com/en-us/office/troubleshoot/excel/determine-column-widths

However, I don't know if I understand your issue. Have you possibility to write a simple test code which currently fails? This issue will be resolved if the test becomes to pass.


---
_I'm contributing ExcelJS just for fun. If want you to know something more go into #1500_

**Comment by fbele** [comment]
> I mean that, which software version are you using to open xlsx (for me it is Microsoft Macintosh Excel v16). It is essential because xlsx format keeps width as a character length, that occurs different width through different OS (dpi changed) or when different default font has been set. More information could be found on the MS Excel documentation page: https://docs.microsoft.com/en-us/office/troubleshoot/excel/determine-column-widths
> 
> However, I don't know if I understand your issue. Have you possibility to write a simple test code which currently fails? This issue will be resolved if the test becomes to pass.
> 
> _I'm contributing ExcelJS just for fun. If want you to know something more go into #1500_

Sorry for the late reply.
Unfortunately I don't know how much of help can I be in this matter as this bug is from long ago. I know that I was using Windows and the latest Excel at that time, probably v16.
Here is my Git repository of the code, but it's at exactly the same state from that time, as I haven't been working on it since, so it's obsolete: https://github.com/fbele/contracts-prototype

Hope it helps!

**Comment by ctheiss** [comment]
I'm also getting this issue; here's some output to help with debugging.

**OS version**
Edition	Windows 10 Home
Version	22H2
Installed on	‎2020-‎07-‎02
OS build	19045.2846
Experience	Windows Feature Experience Pack 120.2212.4190.0

**Environment**
WSL version 2
Ubuntu 18.04.5 LTS
Node v16.19.1
ExcelJS 4.3.0
Excel 2021 MSO (Version 2303 Build 16.0.16227.20202) 64-bit

**Test file**
```js
const ExcelJS = require('exceljs');

const workbook = new ExcelJS.Workbook();
const sheet = workbook.addWorksheet('Test');
sheet.columns = [{ header:'Data', width:10 }];

sheet.addRow([ 50 ]);
sheet.addRow([ 80 ]);

workbook.xlsx.writeFile('test.xlsx');
```

**Output**
![Screenshot 2023-05-01 102219](https://user-images.githubusercontent.com/167414/235468017-1211352d-c2c7-421c-bf48-501bb76d9925.png)

I've also uploaded the resultant file: 
[test.xlsx](https://github.com/exceljs/exceljs/files/11365918/test.xlsx)

**Comment by ctheiss** [comment]
This is perhaps a difference between how the width is specified in the OpenXML and how it's interpreted by Excel.

I found a post on the OpenXML MSDN forum that is very similar to this issue: https://social.msdn.microsoft.com/Forums/sqlserver/en-US/3202fa49-0bae-4846-a1b5-fa29b7f64576/setting-column-width-in-excel?forum=oxmlsdk

This led me to the following explanation of width: https://learn.microsoft.com/en-us/previous-versions/office/developer/office-2010/cc802410(v=office.14)?redirectedfrom=MSDN

> Column width measured as the number of characters of the maximum digit width of the numbers 0, 1, 2, …, 9 as rendered in the normal style's font. There are 4 pixels of margin padding (two on each side), plus 1 pixel padding for the gridlines.
>
>width = Truncate([{Number of Characters} * {Maximum Digit Width} + {5 pixel padding}]/{Maximum Digit Width}*256)/256
>
>[Example: Using the Calibri font as an example, the maximum digit width of 11 point font size is 7 pixels (at 96 dpi). In fact, each digit is the same width for this font. Therefore, if the cell width is 8 characters wide, the value of this attribute must be Truncate([8*7+5]/7*256)/256 = 8.7109375. end example]

So there's the 0.71.

It seems that, in order to fix this issue, we would have to:
1. find all (font, size) pairs used in a column
2. render the digits 0-9 in each pair, and find the max width in pixels
3. add 5/(max width) to the column width

Or, just assume Calibri 11 is used and add 5/7.
